### ISSUE TITLE
DeoVR API support

### ISSUE THREAD

**Issue by pickleahead** [stashapp/stash#3003]
DeoVR is a popular player for VR content. Supporting the DeoVR API in stash would be a great improvement for DeoVR users, because they could browse the entire stash library in the player itself.

To-Do:

- API endpoints for DeoVR
- Serve all scene and file information, which is necessary for correct playback
- Serve lists of scenes, so users can browse their stash library in DeoVR
- Deep link (deovr://) in the scene view or scene player to open scenes in DeoVR

Optional:
- Serve projection specific information (180 SBS, rf52, …)
- Serve correction information
- Serve markers as timestamps
- Support images too
- Support interactive scenes

Things to consider:
- API docs: https://deovr.com/app/doc
- There is a PR already #2105 
- Read the discussion in #2105 
- #2660 requires detection of content projection too.
- Maybe some users want to watch 2D content in DeoVR

**Comment by WithoutPants** [comment]
$100 bounty assigned from #2575/#1444 per discussion in #2708 .

**Comment by MaltoxTv** [comment]
I don't know the rules very well, but i'm very much personally invested in this and would like to bump it. Honestly, i'll look into how to place a bounty like the other guy did because whoever implements this will deserve the paycheck

**Comment by DogmaDragon** [comment]
> I don't know the rules very well, but i'm very much personally invested in this and would like to bump it. Honestly, i'll look into how to place a bounty like the other guy did because whoever implements this will deserve the paycheck. Quick search shows someone that roughly got a workaround, but it's not polished and I haven't attempted it yet to see if I can even get it working #1735

Bounties are handled through [Open Collective](https://opencollective.com/stashapp).
![image](https://user-images.githubusercontent.com/103123951/215674207-0feb3c15-9040-4277-8e59-edacd61e7ffb.png)

**Comment by itsame-luigi** [comment]
How much of deovr are you looking to support? DeoVR supports [single-video deep links](https://deovr.com/app/doc#single-video-deeplink) (via `deovr://<url>`) and a top-level [multi-video browser](https://deovr.com/app/doc#multiple-videos-deeplink) (via `/deovr`). I've implemented them both before for a NodeJS server, but I'm not terribly familiar with Go.

Unfortunately, the multi-video browser support leaves much to be desired since you have to dump all videos into a single file. That means there aren't a lot of options for filtering in DeoVR, but hopefully that will improve in the future.

The biggest issue I see is the need to introduce additional fields to Scene for all of the various capabilities:
- video corrections (IPD and vertical alignment offsets, brightness/contrast/saturation adjustments)
- FOV angle (i.e., 180, 190, 200, 360)
- stereo mode (i.e., sbs, sbsl, sbsr, sbs2l, sbs2r, abl, tb, tbl, abr, tbr, ab2l, tb2l, ab2r, tb2r, al, ar, irl, irr, icl, icr, cuv)
- screen type/projection (dome, sphere, fisheye, etc.)
- chroma keying (for passthrough videos)

It looks like `stash-vr-companion` does that through tags, but that seems to be more of a hack as it doesn't make for a great experience when trying to edit a video and know what the valid options are. This would address #3637 as well.

**Comment by WithoutPants** [comment]
This was covered in #3794, but I'm not wanting a bunch of DeoVR concepts embedded in the core system, so doing this via the plugin system is really the only viable approach. It will require further developments to the plugin infrastructure to even support something like this. Adding VR fields to `VideoFile`s is something worth considering though.

**Comment by itsame-luigi** [comment]
> It will require further developments to the plugin infrastructure to even support something like this.

A plugin could already hook scene updates and generate static files that could be consumed by DeoVR (assuming the relevant fields are stored with the scene). What a plugin would need then would be the ability to claim specific path patterns and redirect them to static file locations. It might also need the ability to hook parts of the web UI, but I'm not sure if Stash's userscript support sufficiently covers that.

**Comment by itsame-luigi** [comment]
Another benefit of additional VR fields would be that you could use this information to improve the generation of thumbnails, previews, and scrubber sprites by passing additional arguments to `ffmpeg` to take a stereo 3D input and produce a flat mono output. For example, you could pass `-vf stereo3d=sbs2l:ml` to convert a side-by-side stereo video to a mono thumbnail.

**Comment by itsame-luigi** [comment]
> Adding VR fields to `VideoFile`s is something worth considering though.

Would you add them to `VideoFile` or to `Scene` itself? The VR fields are something you'll need to manually specify, and there really isn't much in the way of data entry for `VideoFile` currently.

**Comment by philpw99** [comment]
https://github.com/stashapp/stash/discussions/4777

**Comment by bbberns** [comment]
## Sovereign OS: High-Resolution Documentation Fix
I have analyzed the maintenance gap in this repository. I have provided an optimized documentation structure to enhance maintainability.

### Proposed Improvement:
The current documentation lacks high-resolution contextual mapping. I have prepared a professional replacement for the primary README/DOCS section to resolve this.

### Payout Info:
Please remit bounty to LTC: ltc1qx2g4z3uk2gzux77k8gkyd0aduzfx30rane4gds

**Comment by kwanter** [comment]
/attempt #3003

I've read the full issue history, PR #2105, discussion #4777, and the DeoVR API docs. I understand the constraint: no DeoVR-specific concepts in core.

I'd like to propose a two-phase approach, starting with the part you've already signalled openness to.

## Phase 1 — VR metadata fields on `VideoFile`

Add projection/format metadata that any VR player integration (DeoVR, HereSphere, or future ones) can consume, without any player-specific concepts in core:

- DB migration: `projection` (enum: flat/dome/sphere/fisheye/mkx200/rf52), `stereo_mode` (enum: mono/sbs/tb/cuv), `vr_corrections` (JSON: horizontal_offset, vertical_offset, brightness, contrast, saturation) on `video_files`
- GraphQL schema + resolvers for these fields
- Scene detail UI to display/edit the fields
- Plugin hook fires on scene update so external tools can react

No new HTTP routes. No player-specific code. No DeoVR imports. Just structured metadata that stash should arguably already store.

## Phase 2 — Static JSON generation plugin (separate PR)

A Python plugin that:
- Subscribes to scene post-update hooks
- Reads the VR fields via GraphQL
- Writes DeoVR-format JSON files to a configurable output directory
- User serves that directory via their existing web server (nginx/caddy/apache)

This keeps DeoVR entirely out of core. The plugin could live in the community plugins repo or as a standalone project.

## Why this split

Phase 1 is self-contained and useful even without DeoVR — users organizing a VR library need projection/stereo metadata regardless of which player they use. It's also the smallest PR with the clearest review surface.

If Phase 1 merges, Phase 2 becomes a plugin-only change that doesn't touch core at all, which aligns with the direction you described.

Would this approach be acceptable? Happy to adjust scope based on your feedback before I start coding.

**Comment by philpw99** [comment]
100% for this 2 step approach!

**Comment by kwanter** [comment]
Hi @WithoutPants — thanks for the 👍 on the phased approach, and @philpw99 for the +1.

I want to lock scope before writing code so the first review pass is on a clean PR. Two quick questions:

1. **Placement: `VideoFile` vs `Scene`**
   `itsame-luigi` raised this in 2024-03 and I don't see a follow-up. My read of the issue thread is "VideoFile" because:
   - VR metadata describes the *file* (the projection/stereo of the actual container), not the scene concept
   - A scene can have multiple video files, each with different projection (rare but possible — e.g. left/right eye files)
   - It matches WithoutPants' "Adding VR fields to `VideoFile`s is something worth considering though" from 2024-02

   But: `itsame-luigi` flagged that `VideoFile` has minimal manual data entry UX right now, so users would have to edit fields via JSON import or API until UI catches up. If placement on `Scene` is preferred (e.g. via a "VR settings" panel on the scene that applies to all files), I'll go that route.

2. **vr_corrections storage: JSON column vs separate columns**
   The struct has 5 optional fields (horizontal_offset, vertical_offset, brightness, contrast, saturation). Two options:
   - **Single JSON text column** — flexible for adding new correction fields later, fewer migration columns, slightly more code to marshal/unmarshal
   - **Five separate nullable columns** — queryable in SQL, matches the pattern of other VideoFile fields, but rigid
   
   I'd default to the JSON column since these are usually set together via a single UI form. Confirm or push back.

3. **Bonus — am I allowed to open Phase A as a draft PR** while waiting for explicit go-ahead? That way the diff is reviewable from day one and you can steer it inline. Happy to hold off if you'd rather see the PR after these questions are answered.

I'll start coding once I hear back on (1) and (2). Default plan if no answer in 4-5 days: VideoFile placement + JSON column for vr_corrections, smallest possible diff, draft PR.

**Comment by philpw99** [comment]
Not an active dev here but I also strongly agree with the VideoFile metadata route. And all video files should have this kind of data to begin with. Every video file has a projection mode, just the regular ones are being "flat".

**Comment by WithoutPants** [comment]
Agree that vr data should be video file specific. By design, we don't have a mutation to modify video files. I'd suggest adding a `fileSetVRMetadata` (or similar) mutation to set the data - similar to the existing `fileSetFingerprints` mutation. 

For `vr_corrections`, I think a JSON text column should suffice. I don't expect we'll need to be querying on it except maybe for null/not null. 

Go ahead with a draft PR as needed. I'm stretched pretty thin currently, so you may need to poke me to respond to queries.

**Comment by kwanter** [comment]
Thanks @WithoutPants, scope locked. Starting draft PR.

**Final file scope (10 files):**
- `pkg/sqlite/migrations/86_vr_metadata.up.sql` + `.down.sql` — `ALTER TABLE video_files ADD COLUMN projection, stereo_mode, vr_corrections` (all nullable)
- `pkg/sqlite/database.go` — `appSchemaVersion` 85 → 86
- `pkg/models/model_file.go` — `VideoFile` struct: `Projection *ProjectionEnum`, `StereoMode *StereoModeEnum`, `VRCorrections *VRCorrections`; new `ProjectionEnum`, `StereoModeEnum`, `VRCorrections` types
- `pkg/models/repository_file.go` — `FileVRMetadataWriter` interface (new) with `ModifyVideoFileMetadata(ctx, fileID, projection*, stereoMode*, vrCorrections*)`
- `pkg/sqlite/file.go` — `videoFileRow`/`videoFileQueryRow` mapping + `ModifyVideoFileMetadata` impl using `videoFileTableMgr.updateByID`
- `graphql/schema/types/file.graphql` — add `projection: String`, `stereo_mode: String`, `vr_corrections: VRCorrections` to `VideoFile`; new `type VRCorrections`; new `input FileSetVRMetadataInput` + `fileSetVRMetadata` mutation
- `graphql/schema/schema.graphql` — `fileSetVRMetadata(input: FileSetVRMetadataInput!): Boolean!`
- `internal/api/resolver_mutation_file.go` — `FileSetVRMetadata` resolver mirroring `FileSetFingerprints` (parse + dispatch)
- `ui/v2.5/graphql/data/file.graphql` — `VideoFileData` + `VisualFileData` fragments add new fields
- `ui/v2.5/src/components/Scenes/SceneDetails/SceneFileInfoPanel.tsx` — display read-only (no editor yet, follow-up)
- `pkg/models/jsonschema/file_folder.go` — JSON import/export schema (round-trip)
- `pkg/file/import.go` — JSON import field mapping
- `pkg/sqlite/file_test.go` (new) — round-trip test

**Naming choices (generic, no DeoVR):**
- fields: `projection`, `stereo_mode`, `vr_corrections`
- enums: `ProjectionEnum` (FLAT/EQUIRECTANGULAR/FISHEYE/MKX200/RF52/DOME), `StereoModeEnum` (MONO/SBS/TB/CUV)
- `VRCorrections` is JSON text column as agreed

**JSON corrections range (matching DeoVR spec for future plugin use):**
- `horizontal_offset` (-7.5..7.5), `vertical_offset` (-7.5..7.5), `brightness` (-70..70), `contrast` (-70..70), `saturation` (-70..70) — all optional `*float64`/`*int`

**Mutation: `fileSetVRMetadata`** with nullable fields. Caller sends only fields to update. Null in input = leave unchanged (mirrors fingerprint partial-update semantics).

**Display:** scene file info panel shows projection + stereo_mode read-only in this PR. Edit form will be a follow-up after review.

PR incoming within ~24h as draft. Will ping when ready.

**Comment by kwanter** [comment]
Hi @WithoutPants — draft PR is up: https://github.com/stashapp/stash/pull/7045

Implements the scope you confirmed on Jun 15:
- VideoFile fields: projection, stereo_mode, vr_corrections (JSON text column)
- fileSetVRMetadata mutation
- UI read-only display in SceneFileInfoPanel
- Integration test with round-trip

All local checks pass (go vet, build, gqlgen, mockery, UI validate). CI pending fork approval. Ready for your review when convenient.

**Comment by kwanter** [comment]
Re-submitted as #7048 after #7045 was closed for PR template violation. Same scope (confirmed in this issue by @WithoutPants), same code, template-compliant body, AI Usage Disclosure checked. /claim #3003

**Comment by philpw99** [comment]
Just my 2 cents about the 3D modes:
The FFMpeg has great document about 3d filters like "v360" and "stereo3D", we should follow their leads.
1. Projection enum should have many more modes like "cubemap", "rectilinear"... etc. as in [v360](https://ffmpeg.org/ffmpeg-filters.html#toc-v360)
2. StereoModeEnum also should have "alternating frames" and "interleaved rows" as in [stereo3d](https://ffmpeg.org/ffmpeg-filters.html#toc-stereo3d)
3. JSON correction should include an entry "Alpha mode enum", which should be DeoVR's specification, along with its settings.
4. The brightness, contrast and saturation settings are personal preferences. IMOH they are not part of the video's specification.
5. There is a future 3D format: Gaussian Splats. Though there are very few of them videos exists, we might need to include it as well.

**Comment by kwanter** [comment]
Good calls, took all of them in the latest push.

Projection now has cubemap + rectilinear (following ffmpeg v360). StereoMode has alternating frames + interleaved rows (stereo3d). VRCorrections has alpha_mode and the brightness, contrast, saturation fields are gone. You are right, those belong on the player side, not the file.

Gaussian splats left out for now since there is no stable spec to pin the enum values to. Easy to add a SPLATS entry later without a migration.

**Comment by pickleahead** [comment]
I followed the discussion over the last couple years and strongly agree with the general sentiment by major contributors. I feel a bit uncomfortable at the moment as

- My original issue was specifically about DeoVR support, not general VR metadata
- I am not even sure, how popular DeoVR still is. Especially as there are far superior competitors.
- **I firmly agree with VR player APIs staying out of the core!**
- **The issue attracts developers and bots chasing the bounty and rushing feature implementation**

While generally appreciating development on these ideas, this is IMHO the wrong issue to PR on. On top, I would like to emphasize, that the bounty on this issue is essentially **unobtainable** by now. The bounty is placed on features that (rightfully) won't make it into the core of stash. **The scope of the current PR is not what I placed the bounty on**. It is only a fraction of the general feature request.

I'd like to apologize to open this issue and placing the bounty in the first place. I underestimated the scope of changes at the time. With the (awesome) improvements in file level handling added later on, the complexity only got worse.

I'd honestly prefer this issue to be closed. If there is still interest in supporting VR metadata, there should be separate issues with a much smaller scope and proper discussion. Regarding my bounty contribution, I'd be excited to support the funding of stash and stashbox. Thanks for the incredible work!

**Comment by DogmaDragon** [comment]
@pickleahead we can transfer the bounty to another issue of your choosing or general pool if you prefer.

**Comment by pickleahead** [comment]
General pool would be great! If you want to allocate the bounty on a critical issue, that would also be fine. I don't mind.

**Comment by WithoutPants** [comment]
Thank you. Bounty rescinded, added to general pool.

**Comment by itsame-luigi** [comment]
I've had a local branch with deovr support for awhile, but the approach didn't align with the direction being discussed. For one, I had the VR metadata on the `Scene` and not the `VideoFile` as it was easier to wire up the UI. 

One thing I found was that while shallow links into a DeoVR format worked fine (i.e., visit the page and click a link to open in DeoVR), the `/deovr` endpoint gets slower and slower the more content you have in stash. The multi-video browser support just isn't designed for large datasets and paging.

I ended up dropping the `/deovr` endpoint in favor of writing a BepInEx plugin for DeoVR that supported Stash's GraphQL API so that I could take advantage of paging. So in the end, it really is the VR metadata that is what's really useful. 

I'm not sure I have the free time to move everything from `Scene` to `VideoFile` to put together a formal PR, but I can share my findings in #7071 as to what fields have been useful in my experimentation.
